SQL Expressions: Add cell-limit for input dataframes

samjewell opened the pull request:
Replaces https://github.com/grafana/grafana/pull/89252 but takes a similar approach.

A few differences from that previous PR:
- Only makes one `if` check, on the sum total of rows (doesn't try to "fail early")
- Tests the behaviour more completely (should I refactor the tests to be shorter?)
- Doesn't add the `engine` interface or inject any dependency in tests
- Limit Cells, not rows

**What is this feature?**

<img width="500" alt="image" src="https://github.com/user-attachments/assets/7ff054d9-b012-446d-9816-efcf71538f8c" />


Adds a cell-limit to SQL Expressions. This can be configured from the Grafana config file.

**Why do we need this feature?**

Once SQL Expressions is widely available it will be easy for people to filter data within Grafana itself, rather than in the original datasource via the original DS queries they are writing. We want to discourage this, because filtering in Grafana will mean loading much more data into the Grafana instance, which in turn will increase our costs and hurt our reliability - we may see more OOM kills for example.

So we put a limit on the number of cells you can pass to SQL Expressions in order to discourage people from using it in this way.

**Which issue does this fix?**

Not a fix, but part of https://github.com/grafana/grafana-enterprise/issues/7054

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.

samjewell committed:
expr: Add row limit to SQL expressions

Adds a configurable row limit to SQL expressions to prevent memory issues with large
result sets. The limit is configured via the `sql_expression_row_limit` setting in
the `[expressions]` section of grafana.ini, with a default of 100,000 rows.

The limit is enforced by checking the total number of rows across all input tables
before executing the SQL query. If the total exceeds the limit, the query fails
with an error message indicating the limit was exceeded.

asimpson commented on pkg/expr/sql_command_test.go:
```diff
@@ -25,3 +31,115 @@ func TestNewCommand(t *testing.T) {
+	}
+}
+
+func TestExecuteExceedsLimit(t *testing.T) {
```
You could probably pick this test or the one below, I don't think you need both :thinking:

samjewell replied:
Thanks. Single frame cases must always be a subset of multi-frame cases I think - so maybe I'll remove both of the single-frame cases. WDYT?

asimpson commented on pkg/expr/sql_command.go:
```diff
@@ -40,10 +41,13 @@ func NewSQLCommand(refID, rawSQL string) (*SQLCommand, error) {
 		logger.Debug("REF tables", "tables", tables, "sql", rawSQL)
 	}
+
+	defaultLimit := int64(20)
```
Does this line and `expressions.Key("sql_expression_row_limit").MustInt64(100000)` need to be the same?

samjewell replied:
I think I should remove this entirely. Just gonna make sure it all works properly

asimpson reviewed (approved):
In general looks good to me :+1:

samjewell committed:
revert addition of newline

samjewell committed:
Switch to table-driven tests

samjewell committed:
Remove single-frame test-cases.

We only need to test for the multi frame case. Single frame is a subset
of the multi-frame case

samjewell committed:
Add helper function

Simplify the way tests are set up and written

samjewell committed:
Support convention, that limit: 0 is no limit

samjewell committed:
Set the row-limit in one place only

samjewell committed:
Update default limit to 20k rows

As per some discussion here:
https://raintank-corp.slack.com/archives/C071A5XCFST/p1741611647001369?thread_ts=1740047619.804869&cid=C071A5XCFST

samjewell committed:
Test row-limit is applied from config

Make sure we protect this from regressions

This is perhaps a brittle test, somewhat coupled to the code here. But
it's good enough to prevent regressions at least.

samjewell committed:
Add public documentation for the limit

kylebrandt commented on pkg/expr/sql_command.go:
```diff
@@ -121,3 +135,13 @@ func (gr *SQLCommand) Execute(ctx context.Context, now time.Time, vars mathexp.V
 	return TypeSQL.String()
 }
+
+func totalRows(frames []*data.Frame) int64 {
```
Nit, for something this short I would go for a named return val.
```suggestion
func totalRows(frames []*data.Frame) (total int64) {
```

samjewell replied:
I can do this. Can you explain why you prefer it?

kylebrandt replied:
I don't really care, just my impression of what is more "idiomatic", could be wrong though.

samjewell replied:
ok thanks. I made the change anyway :-)

samjewell committed:
Limit total number of cells instead of rows

gabor commented on pkg/expr/nodes.go:
```diff
@@ -103,6 +103,10 @@ func (gn *CMDNode) NeedsVars() []string {
 func (gn *CMDNode) Execute(ctx context.Context, now time.Time, vars mathexp.Vars, s *Service) (mathexp.Results, error) {
+	// If this is a SQL command, pass the cell limit from service config
+	if sqlCmd, ok := gn.Command.(*SQLCommand); ok {
+		sqlCmd.limit = s.cfg.SQLExpressionCellLimit
```
i'm a bit unhappy that we are mutating the sql-command object here.
i would much prefer the `s.cfg.SQLExpressionCellLimit` provided into `NewSQLCommand` as a parameter. that way the sql-expression object is not changing ever, the limit is hardcoded in it.

though i undrstand that at this point we have access to the needed settings-object, so it's a pragmatic choice to do it here.

also there may be some other reason i'm not familiar with (i'm not familiar much with the expression-nodes-system)

still, if possible, sending it into `NewSQLCommand` would be great :)

samjewell replied:
Ok, I've made this change, please see what you think. The code change was a tiny bit more involved, and the test a little longer, but I wouldn't be surprised if you think this is a good change altogether. 

Also I have a question - should we prevent all mutation of these built nodes once they have been built? Definitely not one for this PR, and maybe more of a Golang question than an SSE question, but basically wondering if this can be done, and if people do it in practice?

gabor commented on pkg/expr/sql_command.go:
```diff
@@ -70,18 +72,18 @@ func UnmarshalSQLCommand(rn *rawNode) (*SQLCommand, error) {
 // to execute the command and allows the command to fulfill the Command interface.
-func (gr *SQLCommand) NeedsVars() []string {
-	return gr.varsToQuery
+func (cmd *SQLCommand) NeedsVars() []string {
```
this is a nitpick, but could we do the `gr` => `cmd` renaming in a separate PR?

that would allow this one to be smaller, and the "renaming" PR would be much easier to review and approve.

gabor reviewed (commented):
nice work. in general i think the code looks fine, though i'm not really familiar with how the SSE system works, sorry.

i left two ideas how the situation could be improved. i don't consider them blockers.

kylebrandt commented on pkg/expr/sql_command_test.go:
```diff
@@ -33,16 +34,22 @@ func TestNewCommand(t *testing.T) {
+
+	for c := 0; c < cols; c++ {
+		values := make([]string, rows)
+		for i := range values {
```
Doesn't seem like we are checking the contents of these fields? If that is so, can just leave then zero-valued (empty string)

samjewell replied:
Yeah, the LLM wrote this code, and I had the same observation. I can leave them empty 👍

samjewell committed:
Use named-return for totalRows

As @kylebrandt requested during review of #101700

samjewell committed:
Leave DF cells as zero values during limits tests

When testing the cell limit we don't interact with the cell values at
all, so we leave them at their zero values both to speed up tests, and
to simplify and clarify that their values aren't used.

samjewell committed:
Set SQLCmd limit at object creation - don't mutate

samjewell committed:
Test that SQL node receives limit when built

And that it receives it from the Grafana config

samjewell committed:
Improve TODO message for new Expression Parser

samjewell commented on pkg/expr/service_test.go:
```diff
@@ -146,6 +146,71 @@ func TestDSQueryError(t *testing.T) {
+				RefID:      "A",
+				DataSource: dataSourceModel(),
+				JSON:       json.RawMessage(`{ "datasource": { "uid": "__expr__", "type": "__expr__"}, "type": "sql", "expression": "SELECT 1 AS n" }`),
+				TimeRange: AbsoluteTimeRange{
+					From: time.Time{},
+					To:   time.Time{},
+				},
```
For me this shouldn't be needed, but without this we trigger an error, due to this check:
https://github.com/grafana/grafana/blob/main/pkg/expr/sql_command.go#L52-L55

I might see if I can remove that check completely at some point soon.

samjewell committed:
Fix failing test by always creating config on the Service
